pkg/collector/corechecks/cluster/ksm/kubernetes_state_aggregators.go

```diff
@@ -42,6 +46,28 @@ type countObjectsAggregator struct {
 	counterAggregator
 }
 
+type cronJob struct {
+	namespace string
+	name      string
+}
+
+type cronJobState struct {
+	id    int
+	state metrics.ServiceCheckStatus
+}
+
+type lastCronJobAggregator struct {
+	accumulator map[cronJob]cronJobState
```

Comment:
any reason why this is not a `map[string]cronJobState` where the key is a `ns/name` string?
I can't say using a string as key is better, it's difficult to judge without doing a benchmark.

Reply:
It’s just that I want to index my `cronJobState` objects with two keys and it’s safer use a `struct` than building a string. Think about what would happen if `/` was a valid character in namespace and cron job names.

In addition, I’m not sure there’s a way to build the key string in a way that would be more efficient than building the `struct`.

Here is a small benchmark I did to see the order of magnitude:

```go
package main

import (
	"fmt"
	"testing"
)

var testValues = []struct {
	ns, name string
	value    int
}{
	{"nsA", "nameA", 42},
	{"nsB", "nameB", 43},
	{"nsC", "nameC", 44},
}

func Benchmark_mapStruct(b *testing.B) {

	type mapKey struct {
		namespace string
		name      string
	}

	myMap := make(map[mapKey]int)

	for n := 0; n < b.N; n++ {
		for _, val := range testValues {
			myMap[mapKey{namespace: val.ns, name: val.name}] = val.value
		}
		for _, val := range testValues {
			if v, ok := myMap[mapKey{namespace: val.ns, name: val.name}]; !ok || v != val.value {
				b.Fail()
			}
		}
	}
}

func Benchmark_mapPrintf(b *testing.B) {

	myMap := make(map[string]int)

	for n := 0; n < b.N; n++ {
		for _, val := range testValues {
			myMap[fmt.Sprintf("%s/%s", val.ns, val.name)] = val.value
		}
		for _, val := range testValues {
			if v, ok := myMap[fmt.Sprintf("%s/%s", val.ns, val.name)]; !ok || v != val.value {
				b.Fail()
			}
		}
	}
}

func Benchmark_mapStrConcat(b *testing.B) {

	myMap := make(map[string]int)

	for n := 0; n < b.N; n++ {
		for _, val := range testValues {
			myMap[fmt.Sprintf(val.ns+"/"+val.name)] = val.value
		}
		for _, val := range testValues {
			if v, ok := myMap[fmt.Sprintf(val.ns+"/"+val.name)]; !ok || v != val.value {
				b.Fail()
			}
		}
	}
}
```

```
go test -bench .
goos: darwin
goarch: amd64
pkg: map_struct
cpu: Intel(R) Core(TM) i7-8569U CPU @ 2.80GHz
Benchmark_mapStruct-8      	10415618	       115.1 ns/op
Benchmark_mapPrintf-8      	 1335531	       891.9 ns/op
Benchmark_mapStrConcat-8   	 2215784	       536.2 ns/op
PASS
ok  	map_struct	5.183s
```

Reply:
thanks for taking the time to do the benchmark 🙇